### PR TITLE
Feature: support `StopMarket` `OrderType`

### LINKED ISSUE #7: Adds Support to StopMarketOrder
#### Expected Behavior
Supports `StopMarketOrder`.

#### Actual Behavior
In Lean we have the following message explaining why this implementation doesn't support `StopMarketOrder`:
Despite Binance API allows you to post STOP_LOSS and TAKE_PROFIT order types, they always fail with the content
`{"code":-1013,"msg":"Take profit orders are not supported for this symbol."}`

#### Potential Solution
N/A. Investigate whether this is still the case and why.

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### LINKED ISSUE #38: StopMarket orders are not supported for this symbol : BTCUSDT futures
#### Expected Behavior
it has to submit stop market order without problem, like its described in you api docs

https://www.quantconnect.com/docs/v2/writing-algorithms/trading-and-orders/order-types/stop-market-orders

#### Actual Behavior

```
self.stop_loss = self.stop_market_order(self.symbol , -position_size, stop_loss_price )
 self.take_profit = self.stop_market_order(self.symbol, -position_size, take_profit_price )


04.527 ERROR:: BrokerageModel declared unable to submit order: [1] Warning - Code: NotSupported - StopMarket orders are not supported for this symbol. Please check 'https://api.binance.com/api/v3/exchangeInfo?symbol=BTCUSDT' to see supported
order types.
```

this error dont make sense becouse there is stop market order type 

#### Potential Solution
you can add all possible order types from binance : 

https://api.binance.com/api/v3/exchangeInfo?symbol=BTCUSDT

#### Reproducing the Problem
```
class BasicTemplateCryptoFutureAlgorithm(QCAlgorithm):
    # <summary>
    # Initialise the data and resolution required, as well as the cash and start-end dates for your algorithm. All algorithms must initialized.
    # </summary>

    def initialize(self):
        # # Set start date to 30 days before the current time
        # self.SetStartDate(self.Time - timedelta(days=30))
        # # Set end date for the backtest (can be the current time or any other end date)
        # self.SetEndDate(self.Time)

        # Set Account Currency to Binance Stable Coin for USD
        self.set_account_currency("USDT")
        self.set_cash(200)

        self.usdt_holdings = self.Portfolio.CashBook["USDT"].Amount


        self.set_brokerage_model(BrokerageName.BINANCE_FUTURES, AccountType.MARGIN)

        self.final_closing_trade = None

        self.symbol = "BTCUSDT"

        self.risk_per_trade_pct = 3
        self.taker_fee = 0.0004  # 0.04% taker fee (0.04% per transaction)

        self.add_risk_management(MaximumDrawdownPercentPerSecurity(0.04))

        

        # -------------###################################> tady musi zustat tick, nebo udelat druhy co bude delat tick
        # TOHLE muze vyhodit rate limit ?

        self.crypto_future = self.add_crypto_future(self.symbol, Resolution.TICK)
        # perpetual futures does not have a filter function

        # self.crypto_future.set_buying_power_model(CryptoFutureMarginModel( 20 ))


        self.interest_per_symbol = {self.symbol:  0, }


        # the amount of BTC we need to hold to trade 'BTCUSD'
        # self.btc_usd.base_currency.set_amount(0.005)


        # ==========================================================
        self.btc_usd_holdings = 0
        # Every day symbol is trading, call the 'set_candle_ready' method every 30 minutes.
        self.is_candle_ready = False

        self.schedule.on(
            self.date_rules.every_day(self.symbol),
            self.time_rules.every(timedelta(seconds = 10 )),
            self.set_candle_ready
        )


        self.api_binance = API_binance()

        self.set_time_zone(TimeZones.UTC)
        #
        self.strat_prod = Strat_prod()
        #


    def set_candle_ready(self):
        self.is_candle_ready = True



    def get_crypto_quantity(self, symbol, percentage, stop_loss_price):

        trade_price = self.crypto_future.bid_price

        crypto = self.securities[self.symbol]

        position_size = (self.usdt_holdings * (self.risk_per_trade_pct / 100)) / (trade_price - stop_loss_price )



        # # # Round down to observe the lot size
        # lot_size = crypto.symbol_properties.lot_size
        # quantity = round( position_size / lot_size) * lot_size


        # self.log(f"quantity side of type: { quantity } with target_quantity: {target_quantity}")

        self.log(f" position_size : { position_size} with target_quantity: {position_size *  trade_price }")

        # if self.is_valid_order_size(crypto, quantity):
        #     self.market_order(symbol, quantity)
        return position_size

    # Brokerages have different order size rules
    # Binance considers the minimum volume (price x quantity):
    def is_valid_order_size(self, crypto, quantity):
        return abs(crypto.price * quantity) > crypto.symbol_properties.minimum_order_size




    # <summary>
    # OnData event is the primary entry point for your algorithm. Each new data point will be pumped in here.
    # </summary>
    # <param name="data">Slice object keyed by symbol containing the stock data</param>
    def OnData(self, slice):

        interest_rates = slice.Get(MarginInterestRate)
        for interest_rate in interest_rates:
            self.interest_per_symbol[interest_rate.key] += 1
            self.cached_interest_rate = self.securities[interest_rate.key].cache.get_data[MarginInterestRate]()

            if self.cached_interest_rate != interest_rate.value:
                raise Exception(f"Unexpected cached margin interest rate for {interest_rate.key}!")




# ======================================================================
        self.usdt_holdings = self.Portfolio.CashBook["USDT"].Amount


        if(self.is_candle_ready ):
            # and not df_ohlcv.empty
            self.is_candle_ready = False

            df_ohlcv = self.api_binance.get_klines_historical()

            # self.log(df_ohlcv.head(1))
            # self.log(df_ohlcv.tail(1))
            # self.log(" ================================================= ")

            # last_row, df_signal = self.strat_prod.run_test( df_ohlcv)
            # self.log(df_signal.describe() )



            #
            if ( True ):

                # --------##################### TOHLE BUDE DELAT JENOM SINGLE TRADE
                #                               PRO TESTING TO JE ASI V POHODE
                if self.portfolio.invested == False and self.transactions.orders_count == 0:

                    # order_properties = OrderProperties()
                    # self.log(f"self.order_properties : { dir(order_properties)}")
                    # order_properties.type = "STOP_LOSS"


                    # TADY JE BUY ---> jaky buy kdyz se tam dela sell? last_row.SL.values

                    entry_price = self.crypto_future.bid_price

                    stop_loss_price = entry_price * 0.995  # Example stop loss at 2% below entry
                    take_profit_price = entry_price * 1.005


                    info = self.client_binance.exchange_info()
                # 'pricePrecision': 2, 'quantityPrecision': 3, 'baseAssetPrecision': 8, 'quotePrecision': 8,



                    # position_size = self.get_position_size(entry_price=entry_price, stop_loss_price=  stop_loss_price , side='BUY' )

                    position_size = self.get_crypto_quantity(self.symbol, percentage= 0.03, stop_loss_price= stop_loss_price)



                    # Place market order
                    # ticket = self.buy(self.symbol, position_size)

                    # self.debug(f"Quantity filled: {ticket.quantity_filled}; Fill price: {ticket.average_fill_price}")
                    position_size = float(round(position_size, 2))
                    stop_loss_price = float(round(stop_loss_price, 2))
                    take_profit_price = float(round(take_profit_price, 2))

                    self.stop_loss = self.stop_market_order(self.symbol , -position_size, stop_loss_price )
                    self.take_profit = self.stop_market_order(self.symbol, -position_size, take_profit_price )






                    self.debug(f"stop_loss_price ; take_profit_price : {stop_loss_price, take_profit_price }; position_size; entry_price : { position_size, entry_price }")

```
#### System Information
ubuntu 22.04 LTS

#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by Romazes**
#### Description

- Add supporting of `StopMarket` Order Type for Futures implementation.
- Fix `GetAccountHoldings()` && `GetCashBalance()` in `class BinanceCoinFuturesRestApiClient`
  - Actual Behavior: when we have tried to get `GetAccountHolding()` or `GetCashBalance()`, we sent request to BinanceFuture endpoint instead of BinanceFutureCoin.
  - Old realization uses the wrong API prefix of **BinanceFuture**: `/fapi/` instead of  **BinanceCoinFutures**: `/dapi`.

##### Exchange Info
Validate supporting of OrderTypes for specific Trading pair:
- [SPOT](https://api.binance.com/api/v3/exchangeInfo)
- [Futures](https://fapi.binance.com/fapi/v1/exchangeInfo)
- [FuturesCoin](https://dapi.binance.com/dapi/v1/exchangeInfo)

image

##### Support Answer 
The screenshot links:
- https://developers.binance.com/docs/binance-spot-api-docs/rest-api/public-api-endpoints#new-order-trade
- https://developers.binance.com/docs/derivatives/usds-margined-futures/trade/rest-api

image

#### Related PRs
- https://github.com/QuantConnect/Lean/pull/8441

#### Related Issue

closes #7 
closes #38 

#### Motivation and Context

Supporting additional order types to user can create new algorithms with this type.

#### Requires Documentation Change

N/A

#### How Has This Been Tested?

- run a specific test with `StopMarket` OrderType in Binance-CointFuture||FuturesBrokerageTests.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Review by Martin-Molinero — APPROVED** [review]
Nice! Thank you 👍

**Comment by lukaspistelak** [comment]
so it will be works now stop market and take profit market order ?

**Comment by Romazes** [comment]
> so it will be works now stop market and take profit market order ?

Hey, man
- `StopMarket` works with `BinanceFuture` and `BinanceFutureCoin`.
- [Take Profit Limit](https://github.com/QuantConnect/Lean.Brokerages.Binance/blob/57e9b15029f909e3d43c9856e53295b2a0d2ca22/QuantConnect.BinanceBrokerage/BinanceBaseRestApiClient.cs#L295)
- we don't support `Take Profit Market`

**Comment by lukaspistelak** [comment]
> > so it will be works now stop market and take profit market order ?
> 
> Hey, man
> 
>     * `StopMarket` works with `BinanceFuture` and `BinanceFutureCoin`.
> 
>     * [Take Profit Limit](https://github.com/QuantConnect/Lean.Brokerages.Binance/blob/57e9b15029f909e3d43c9856e53295b2a0d2ca22/QuantConnect.BinanceBrokerage/BinanceBaseRestApiClient.cs#L295)
> 
>     * we don't support `Take Profit Market`

ok, so take profit limit works? how this works exactly?  is this standart limit order? limit orders are absolute useless for TP and SL  ( you have to solve many problems with partial execution of order and etc.....)

**Comment by Romazes** [comment]
> > > so it will be works now stop market and take profit market order ?
> > 
> > 
> > Hey, man
> > ```
> > * `StopMarket` works with `BinanceFuture` and `BinanceFutureCoin`.
> > 
> > * [Take Profit Limit](https://github.com/QuantConnect/Lean.Brokerages.Binance/blob/57e9b15029f909e3d43c9856e53295b2a0d2ca22/QuantConnect.BinanceBrokerage/BinanceBaseRestApiClient.cs#L295)
> > 
> > * we don't support `Take Profit Market`
> > ```
> 
> ok, so take profit limit works? how this works exactly? is this standart limit order? limit orders are absolute useless for TP and SL ( you have to solve many problems with partial execution of order and etc.....)

To implement a Take Profit or Stop Loss order in Lean, you should use the `StopLimitOrder` type. The logic depends on the `Stop Price` relative to the current ticker price to determine whether the order is a `"STOP_LOSS_LIMIT"` or `"TAKE_PROFIT_LIMIT"`.

```
if (order.Direction == OrderDirection.Sell)
{
    body["type"] = stopPrice <= ticker ? "STOP_LOSS_LIMIT" : "TAKE_PROFIT_LIMIT";
}
else
{
    body["type"] = stopPrice <= ticker ? "TAKE_PROFIT_LIMIT" : "STOP_LOSS_LIMIT";
}
```

> please join to us discord channel and create [ticket](https://www.quantconnect.com/docs/v2/cloud-platform/community/discord)

**Comment by lukaspistelak** [comment]
> > > > so it will be works now stop market and take profit market order ?
> > > 
> > > 
> > > Hey, man
> > > ```
> > > * `StopMarket` works with `BinanceFuture` and `BinanceFutureCoin`.
> > > 
> > > * [Take Profit Limit](https://github.com/QuantConnect/Lean.Brokerages.Binance/blob/57e9b15029f909e3d43c9856e53295b2a0d2ca22/QuantConnect.BinanceBrokerage/BinanceBaseRestApiClient.cs#L295)
> > > 
> > > * we don't support `Take Profit Market`
> > > ```
> > 
> > 
> > ok, so take profit limit works? how this works exactly? is this standart limit order? limit orders are absolute useless for TP and SL ( you have to solve many problems with partial execution of order and etc.....)
> 
> To implement a Take Profit or Stop Loss order in Lean, you should use the `StopLimitOrder` type. The logic depends on the `Stop Price` relative to the current ticker price to determine whether the order is a `"STOP_LOSS_LIMIT"` or `"TAKE_PROFIT_LIMIT"`.
> 
> ```
> if (order.Direction == OrderDirection.Sell)
> {
>     body["type"] = stopPrice <= ticker ? "STOP_LOSS_LIMIT" : "TAKE_PROFIT_LIMIT";
> }
> else
> {
>     body["type"] = stopPrice <= ticker ? "TAKE_PROFIT_LIMIT" : "STOP_LOSS_LIMIT";
> }
> ```
> 
> > please join to us discord channel and create [ticket](https://www.quantconnect.com/docs/v2/cloud-platform/community/discord)

ok ,when it will be in new release in pip lib ?
